### PR TITLE
Provide service accounts tokens to extensions

### PR DESCRIPTION
**Pull request by scrawfor99**
### Description
This PR implements service account issuance and passing for extensions. This change adds a new transport action which passes the extension a string representation of its service account auth token. This token is created by the TokenManager interface implementation. The token is expected to be an encoded basic auth credential string which can be used by the extension to interact with its own system index. 

As part of this change, the ExtensionManager will now take an instance of the IdentityService as part of its construction. This change was made so that the service account token could be generated inside of the class using a call to the token manager. Without passing the IdentityService it was not possible to get the implementation of the TokenManager required to generate the ServiceAccountToken. 

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

This issue resolves: https://github.com/opensearch-project/security/issues/3176

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/node/Node.java
Comment (cwperks) [inline server/src/main/java/org/opensearch/node/Node.java:487]: I think the constructor is the way to go, but I do see that ExtensionsManager has setters (albeit unused) for other services like this: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/extensions/ExtensionsManager.java#L490-L492

File: server/src/main/java/org/opensearch/extensions/action/IssueServiceAccountResponse.java
Comment (cwperks) [inline server/src/main/java/org/opensearch/extensions/action/IssueServiceAccountResponse.java:24]: nit: avoid String in variable names. The datatype gives it away that this is a string. 

Since this is the response object, what is the value of including the token on the response?

Comment (scrawfor99) [reply server/src/main/java/org/opensearch/extensions/action/IssueServiceAccountResponse.java:24]: Oh shoot, yeah I don't know what I was thinking for the variable name....

For the response, I was thinking it would be good to have based on the response objects from what I saw in the `InitializeExtensionResponse` file where there is this:


```
    private String name;
    private List<String> implementedInterfaces;

    public InitializeExtensionResponse(String name, List<String> implementedInterfaces) {
        this.name = name;
        this.implementedInterfaces = implementedInterfaces;
    }

    public InitializeExtensionResponse(StreamInput in) throws IOException {
        name = in.readString();
        this.implementedInterfaces = Arrays.asList(in.readStringArray());
    }

    @Override
    public void writeTo(StreamOutput out) throws IOException {
        out.writeString(name);
        out.writeStringArray(implementedInterfaces.toArray(new String[0]));
    }

    /**
     * @return the node that is currently leading, according to the responding node.
     */

    public String getName() {
        return this.name;
    }

    /**
     * @return interfaces implemented by an extension
     */

    public List<String> getImplementedInterfaces() {
        return implementedInterfaces;
    }

    @Override
    public String toString() {
        return "InitializeExtensionResponse{" + "name = " + name + " , " + "implementedInterfaces = " + implementedInterfaces + "}";
    }

    @Override
```
It seemed like the authors were being explicit as to what extension was made and with what interfaces so I figured I would provide explicit info as well.

I can remove it if you think that is better.

**Comment by scrawfor99** [comment]
Flaky test:
```
[org.opensearch.index.shard.IndexShardTests.testRefreshLevelRestoreShardFromRemoteStore](https://build.ci.opensearch.org/job/gradle-check/23752/testReport/junit/org.opensearch.index.shard/IndexShardTests/testRefreshLevelRestoreShardFromRemoteStore/)
```

File: server/src/main/java/org/opensearch/extensions/action/IssueServiceAccountResponse.java
Comment (cwperks) [reply server/src/main/java/org/opensearch/extensions/action/IssueServiceAccountResponse.java]: What would be useful for the response? Can the extension report back if the operation was successful or not? What happens if the transport action times out? (This scenario doesn't need to be handled in this PR, but could represent a real scenario) Would the action be retried if unsuccessful?

Comment (scrawfor99) [reply server/src/main/java/org/opensearch/extensions/action/IssueServiceAccountResponse.java]: I think we just need to know that the extension received the service account token. I am not sure that anything other than that is required. 

For retrying and time outs, I would expect extension installation to fail probably but that will likely be a separate PR if we want something more sophisticated than just throwing the same type of error that the InitializeExtensionAction does.

**Comment by scrawfor99** [comment]
Flaky tests: 

```


  | Test Result (2 failures  / +2)org.opensearch.cluster.allocation.ClusterRerouteIT.testDelayWithALargeAmountOfShardsorg.opensearch.remotestore.RemoteIndexRecoveryIT.testRerouteRecovery
-- | --


[Test Result](https://build.ci.opensearch.org/job/gradle-check/23981/testReport/) (2 failures / +2)

    [org.opensearch.cluster.allocation.ClusterRerouteIT.testDelayWithALargeAmountOfShards](https://build.ci.opensearch.org/job/gradle-check/23981/testReport/junit/org.opensearch.cluster.allocation/ClusterRerouteIT/testDelayWithALargeAmountOfShards/)
    [org.opensearch.remotestore.RemoteIndexRecoveryIT.testRerouteRecovery](https://build.ci.opensearch.org/job/gradle-check/23981/testReport/junit/org.opensearch.remotestore/RemoteIndexRecoveryIT/testRerouteRecovery/)
```

**Comment by scrawfor99** [comment]
Can this please be reviewed? Thank you!

File: server/src/main/java/org/opensearch/node/Node.java
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/node/Node.java:497]: FYI, the setters were (are still?) there because of the order in which we initiated, a chicken-and-egg sort of a thing.  I agree we shouldn't use them if we don't need them.

File: server/src/test/java/org/opensearch/extensions/action/IssueServiceAccountRequestTests.java
Comment (dbwiddis) [inline server/src/test/java/org/opensearch/extensions/action/IssueServiceAccountRequestTests.java:35]: (Personal preference, don't need to fix) I always thought it was silly to test exact format of a toString instead of just checking that it contains the important info.

File: plugins/identity-shiro/src/main/java/org/opensearch/identity/shiro/ShiroTokenManager.java
Comment (dbwiddis) [inline plugins/identity-shiro/src/main/java/org/opensearch/identity/shiro/ShiroTokenManager.java:82]: Elsewhere in the code (I think [UUID generation](https://github.com/opensearch-project/OpenSearch/blob/main/libs/common/src/main/java/org/opensearch/common/RandomBasedUUIDGenerator.java#L60C19-L60C19)) we've used URL-safe encoders.  Since this will be passed in headers I think that's needed here (and possibly when decoding?)

**Review by dbwiddis — COMMENTED** [review]
One substantive comment.

File: plugins/identity-shiro/src/main/java/org/opensearch/identity/shiro/ShiroTokenManager.java
Comment (scrawfor99) [reply plugins/identity-shiro/src/main/java/org/opensearch/identity/shiro/ShiroTokenManager.java:82]: Oh okay, I will make the swap.

File: server/src/test/java/org/opensearch/extensions/action/IssueServiceAccountRequestTests.java
Comment (scrawfor99) [reply server/src/test/java/org/opensearch/extensions/action/IssueServiceAccountRequestTests.java:35]: I will keep that in mind! Thank you :)

**Review by dbwiddis — APPROVED** [review]
LGTM!

**Comment by scrawfor99** [comment]
Unrelated test failure:

```
[org.opensearch.index.translog.RemoteFsTranslogTests.testConcurrentWritesWithVaryingSize](https://build.ci.opensearch.org/job/gradle-check/25605/testReport/junit/org.opensearch.index.translog/RemoteFsTranslogTests/testConcurrentWritesWithVaryingSize/)
```

**Review by owaiskazi19 — COMMENTED** [review]
Small comments! Rest LGTM!

File: server/src/main/java/org/opensearch/extensions/ExtensionsManager.java
Comment (scrawfor99) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:459]: I am not 100% I understand your question, but I think you are asking why have it a string? This converts it into a standardized format for sending over REST requests.

File: server/src/main/java/org/opensearch/extensions/rest/RestInitializeExtensionAction.java
Comment (scrawfor99) [reply server/src/main/java/org/opensearch/extensions/rest/RestInitializeExtensionAction.java:164]: It is separate to keep the transport actions isolated from one another.

File: server/src/main/java/org/opensearch/extensions/action/IssueServiceAccountResponse.java
Comment (scrawfor99) [reply server/src/main/java/org/opensearch/extensions/action/IssueServiceAccountResponse.java:65]: I will remove it

File: plugins/identity-shiro/src/main/java/org/opensearch/identity/shiro/ShiroTokenManager.java
Comment (scrawfor99) [reply plugins/identity-shiro/src/main/java/org/opensearch/identity/shiro/ShiroTokenManager.java:85]: This is just a test class but we can revisit some constants down the road if that works.

File: plugins/identity-shiro/src/main/java/org/opensearch/identity/shiro/ShiroTokenManager.java
Comment (scrawfor99) [reply plugins/identity-shiro/src/main/java/org/opensearch/identity/shiro/ShiroTokenManager.java:83]: Oops.

File: server/src/main/java/org/opensearch/extensions/ExtensionsManager.java
Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:459]: I meant can we have the variable name just as `authToken`?

**Review by owaiskazi19 — APPROVED** [review]
Won't hold the merge. LGTM!

File: server/src/main/java/org/opensearch/extensions/rest/RestInitializeExtensionAction.java
Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/extensions/rest/RestInitializeExtensionAction.java:164]: But won't we have to issue a service account every time we are initializing an extension?

**Comment by scrawfor99** [comment]
```


  | Test Result (6 failures  / +5)org.opensearch.indices.replication.SegmentReplicationSuiteIT.testFullRestartDuringReplicationorg.opensearch.indices.replication.SegmentReplicationSuiteIT.testDeleteIndexWhileReplicatingorg.opensearch.indices.replication.SegmentReplicationSuiteIT.testBasicReplicationorg.opensearch.indices.replication.SegmentReplicationSuiteIT.classMethodorg.opensearch.remotestore.RemoteStoreStatsIT.testDownloadStatsCorrectnessSinglePrimarySingleReplicaorg.opensearch.remotestore.SegmentReplicationUsingRemoteStoreIT.testUpdateOperations
-- | --


[Test Result](https://build.ci.opensearch.org/job/gradle-check/25654/testReport/) (6 failures / +5)

    [org.opensearch.indices.replication.SegmentReplicationSuiteIT.testFullRestartDuringReplication](https://build.ci.opensearch.org/job/gradle-check/25654/testReport/junit/org.opensearch.indices.replication/SegmentReplicationSuiteIT/testFullRestartDuringReplication/)
    [org.opensearch.indices.replication.SegmentReplicationSuiteIT.testDeleteIndexWhileReplicating](https://build.ci.opensearch.org/job/gradle-check/25654/testReport/junit/org.opensearch.indices.replication/SegmentReplicationSuiteIT/testDeleteIndexWhileReplicating/)
    [org.opensearch.indices.replication.SegmentReplicationSuiteIT.testBasicReplication](https://build.ci.opensearch.org/job/gradle-check/25654/testReport/junit/org.opensearch.indices.replication/SegmentReplicationSuiteIT/testBasicReplication/)
    [org.opensearch.indices.replication.SegmentReplicationSuiteIT.classMethod](https://build.ci.opensearch.org/job/gradle-check/25654/testReport/junit/org.opensearch.indices.replication/SegmentReplicationSuiteIT/classMethod/)
    [org.opensearch.remotestore.RemoteStoreStatsIT.testDownloadStatsCorrectnessSinglePrimarySingleReplica](https://build.ci.opensearch.org/job/gradle-check/25654/testReport/junit/org.opensearch.remotestore/RemoteStoreStatsIT/testDownloadStatsCorrectnessSinglePrimarySingleReplica/)
    [org.opensearch.remotestore.SegmentReplicationUsingRemoteStoreIT.testUpdateOperations](https://build.ci.opensearch.org/job/gradle-check/25654/testReport/junit/org.opensearch.remotestore/SegmentReplicationUsingRemoteStoreIT/testUpdateOperations/)
```
Flaky tests

**Comment by scrawfor99** [comment]
Hi @dbwiddis could you merge this when you get a chance? Thanks.

File: server/src/main/java/org/opensearch/extensions/ExtensionsManager.java
Comment (peternied) [inline server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:460]: This object it isn't passed into any methods to be waited on, so it seems like we can remove it?

Comment (peternied) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:460]: Note; I think this is still the case in the initialization flow and it was copy/pasted here.

File: server/src/main/java/org/opensearch/extensions/action/IssueServiceAccountResponse.java
Comment (peternied) [inline server/src/main/java/org/opensearch/extensions/action/IssueServiceAccountResponse.java:24]: Why does the response need to be anything more than an acknowledgement?  The extension either is setup with the account or it doesn't need it / consider it.

**Review by peternied — CHANGES_REQUESTED** [review]
Thanks for wiring this up; however, I think the data model for the extension is made less clear with this change.  I think extensions should have a single state 'blob' from OpenSearch, handing initialization and updates in a single place.  Otherwise as we add new bits of functionality we are adding new request handlers for each 'field'.

As @owaiskazi19 points out, the initialization request seems like the right place to add this information passing - not a new request.  Since extensions are still experimental we've got flexibility to build up and out this system if needed, for this scenario I think the less moving parts (less requests) the easier it will be to handle.

File: server/src/main/java/org/opensearch/extensions/ExtensionsManager.java
Comment (scrawfor99) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:456]: Craig asked it to be  separate transport request:  https://github.com/opensearch-project/security/issues/3176.

Comment (cwperks) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:456]: @peternied What if a token needs to be re-issued? I'm in favor of separate transport request

File: server/src/main/java/org/opensearch/extensions/action/IssueServiceAccountResponse.java
Comment (scrawfor99) [reply server/src/main/java/org/opensearch/extensions/action/IssueServiceAccountResponse.java:24]: It doesn't, I just put this information for verifiability etc. Felt it could be good for audit logging. Want me to remove it?

File: server/src/main/java/org/opensearch/extensions/ExtensionsManager.java
Comment (scrawfor99) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:460]: Yeah will do. I just copied from the other flow, but I will remove it.

File: server/src/main/java/org/opensearch/extensions/ExtensionsManager.java
Comment (peternied) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:456]: Then the state should be updated, retransmitted, in a world where permissions of the extension change there should be a single kind of notification where the extension can know what it can/cannot do.  As I said we can always build it out later, but lets start small as this is an experimental feature.

Comment (cwperks) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:456]: I agree that it should be more generic. `initializeExtensionSecuritySettings`?

In a draft PR I have open in the sdk there's a linked PR to my fork in core that shows a transport action called `initializeExtensionSecurity`.

https://github.com/cwperks/OpenSearch/pull/94/files#diff-c9a4accc046eb92a0e6a025970b3eacacc330f40bb55b02859df05a79710c8c8R420

Comment (scrawfor99) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:456]: @peternied @cwperks what is the call?

Comment (scrawfor99) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:456]: I will make it part of initialize like Peter and Owais suggested.

File: server/src/main/java/org/opensearch/extensions/ExtensionsManager.java
Comment (scrawfor99) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:459]: Gotcha. fixed.

File: server/src/main/java/org/opensearch/extensions/action/IssueServiceAccountResponse.java
Comment (scrawfor99) [reply server/src/main/java/org/opensearch/extensions/action/IssueServiceAccountResponse.java:24]: Removed so all set.

File: server/src/main/java/org/opensearch/extensions/rest/RestInitializeExtensionAction.java
Comment (scrawfor99) [reply server/src/main/java/org/opensearch/extensions/rest/RestInitializeExtensionAction.java:164]: Will do.

**Comment by scrawfor99** [comment]
@peternied what do you think about this now?

File: server/src/main/java/org/opensearch/extensions/ExtensionsManager.java
Comment (peternied) [inline server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:454]: Remove public modifier, external classes shouldn't use this IMO

File: server/src/main/java/org/opensearch/node/Node.java
Comment (peternied) [reply server/src/main/java/org/opensearch/node/Node.java:497]: This seems out of scope for the current change as it isn't making the problem _worse_

File: server/src/main/java/org/opensearch/extensions/ExtensionsManager.java
Comment (peternied) [inline server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:80]: Remove?

File: CHANGELOG.md
Comment (peternied) [inline CHANGELOG.md:15]: Remove these entries?

**Review by peternied — COMMENTED** [review]
Just minor cleanup, then should be ready to merge - much smaller change good work @scrawfor99

File: server/src/main/java/org/opensearch/extensions/ExtensionsManager.java
Comment (scrawfor99) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:454]: Good catch

File: server/src/main/java/org/opensearch/extensions/ExtensionsManager.java
Comment (scrawfor99) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:80]: Oops, forgot to remove that.

File: CHANGELOG.md
Comment (scrawfor99) [reply CHANGELOG.md:15]: Oops will do

**Review by peternied — APPROVED** [review]

**Comment by peternied** [comment]
@scrawfor99 Waiting on CI to complete since merge conflicts were resolved

**Comment by scrawfor99** [comment]
@peternied should be all set :D

**Comment by peternied** [comment]
@scrawfor99 Could you make sure all the flaky test have issues or file new ones, this was a rough go of things

**Comment by scrawfor99** [comment]
Will do

**Pull request merged**
